backend/src/main/scala/bloop/Compiler.scala

```diff
@@ -386,14 +368,11 @@ object Compiler {
         logger,
         uniqueInputs,
         newFileManager,
-        cancelPromise,
         tracer,
         classpathOptions
       )
       .materialize
-      .doOnCancel(Task(cancel()))
       .map {
-        case Success(_) if cancelPromise.isCompleted => handleCancellation
```

Comment:
@dos65 Could you double check my logic?

If cancel promise is completed in the doOnCancel above, which is the only place it's completed, then we would never get to this map unless we already went through it.

In other words, would this branch would ever be achievable?

Reply:
The reason why I am thinking this is that if cancelation cancel a particular task we don't expect it to go through the next chains, right? We wanted all the cancels to happen, but not really anything else.

Reply:
I might need to dig in here a bit more. I haven't noticed other usage.

Reply:
> If cancel promise is completed in the doOnCancel above, which is the only place it's completed, then we would never get to this map unless we already went through it

@tgodzik No, we always goes trough all task computation. The `cancel` in monix2 only performs lamdas passed in `doOnCancel`. In this case this removed `handleCancellation` is actually reachable and valid

Reply:
Ach, right so that is why this is not really cancelation :thinking: 

I remember you had some small experiments with Monix, do you have it still somewhere?

Reply:
😄 what about cancellation, it was just small samples - I think I placed them in tests: `cancel1`, `cancel2` https://github.com/scalacenter/bloop/blob/fa0d88309eb88aea86376f6613d5923aaaced2dc/backend/src/test/scala/bloop/task/TaskSpec.scala#L24

Reply:
If we do cancelation this way I wonder if we shouldn't just use Future :thinking: and bsp4j library, where we could handle things via completable futures + promises. Is there an actual benefit of using current monix cancelation in comparison to just Future?

I wonder if rewriting to cats will just make us use another library not really suited for out purposes.

Reply:
From the perspective of getting rid of monix, I don't think that there will be major difference in choice between cats-io vs delayed version of future `() => Future`. The amount of work will be similar.

Right now, we are using an embedded cancelation mechanic in monix2, while it would be better to separate that thing.
I mean that instead of cancelling a whole bsp-request computation, what we need is to interrupt a compilation + maybe test/run tasks which would fit to any async data-type.
I mean:
```scala
// 
trait BloopTask[A] {
  def result: F[A]
  def cancel: F[Unit]
}

object Compiler {
  //  Return a task that return a specific bloop task (for example compile)
  //  then at CLI/BSP side if we want to cancel, we cancel only long-running operations that are designed to be interrupted 
  def compile(...): Task[BloopTask[Result]] = {
     ...
  }
}
```
However,  I have no idea how much time it will take.